app/controllers/financialOverview/PsaPenaltiesAndChargeDetailsController.scala

```diff
@@ -0,0 +1,144 @@
+import play.api.i18n.{I18nSupport, Messages, MessagesApi}
+import play.api.libs.json.{JsObject, Json}
+import play.api.mvc._
+import renderer.Renderer
+import services.SchemeService
+import services.financialOverview.PsaPenaltiesAndChargesService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import uk.gov.hmrc.viewmodels.{Html, NunjucksSupport}
+import utils.DateHelper.dateFormatterDMY
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class PsaPenaltiesAndChargeDetailsController @Inject()(identify: IdentifierAction,
+                                                        allowAccess: AllowAccessActionProviderForIdentifierRequest,
+                                                        override val messagesApi: MessagesApi,
+                                                        val controllerComponents: MessagesControllerComponents,
+                                                        psaPenaltiesAndChargesService: PsaPenaltiesAndChargesService,
+                                                        schemeService: SchemeService,
+                                                        renderer: Renderer
+                                                      )(implicit ec: ExecutionContext)
+  extends FrontendBaseController
+    with I18nSupport
+    with NunjucksSupport {
+
+  def onPageLoad(identifier: String,
+                 chargeReferenceIndex: String,
+                 journeyType: ChargeDetailsFilter): Action[AnyContent] =
+    (identify andThen allowAccess()).async {
+    implicit request =>
+      psaPenaltiesAndChargesService.getPenaltiesForJourney(request.idOrException, journeyType).flatMap { penaltiesCache =>
+
+        val chargeRefs: Seq[String] = penaltiesCache.penalties.map(_.chargeReference)
+        def penaltyOpt: Option[PsaFSDetail] = penaltiesCache.penalties.find(_.chargeReference == chargeRefs(chargeReferenceIndex.toInt))
+
+        if(chargeRefs.length > chargeReferenceIndex.toInt && penaltyOpt.nonEmpty) {
+          schemeService.retrieveSchemeDetails(request.idOrException, identifier, "pstr") flatMap {
+            schemeDetails =>
+              val json = Json.obj(
+                "psaName" -> penaltiesCache.psaName,
+                "schemeAssociated" -> true,
+                "schemeName" -> schemeDetails.schemeName
+              ) ++ commonJson(penaltyOpt.head, penaltiesCache.penalties, chargeRefs, chargeReferenceIndex, journeyType)
+
+              renderer.render(template = "financialOverview/psaChargeDetails.njk", json).map(Ok(_))
+            }
+          } else {
+          Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad))
+        }
+      }
+  }
+
+  private def setInsetText(psaFS: PsaFSDetail, interestUrl: String, originalChargeUrl: String)
+                           (implicit messages: Messages): Html = {
+    if (psaFS.chargeType == CONTRACT_SETTLEMENT && psaFS.accruedInterestTotal > 0) {
```

Comment:
doing get on option cloud lead to an exception if the return value is None, I think it is better to check if (isDefined) before calling get on Option value